File: android/app/src/main/cpp/streamnative/plugins/StreamXmlPlugin.cpp
#include "StreamXmlPlugin.h"

namespace streamnative {

StreamXmlPlugin::StreamXmlPlugin(bool includeTagsInOutput)
        : includeTagsInOutput_(includeTagsInOutput),
          state_(PluginState::IDLE),
          startState_(StartState::WAIT_LT),
          allowStartAfterEndTag_(false),
          allowStartAfterPunctuation_(false),
          haveEndPattern_(false) {
    reset();
}

PluginState StreamXmlPlugin::state() const {
    return state_;
}

bool StreamXmlPlugin::initPlugin() {
    reset();
    return true;
}

void StreamXmlPlugin::reset() {
    state_ = PluginState::IDLE;
    startState_ = StartState::WAIT_LT;
    tagName_.clear();
    endMatcher_.reset();
    endPattern_.clear();
    haveEndPattern_ = false;
    lastChar_ = 0;
}

bool StreamXmlPlugin::isAsciiLetter(char16_t c) {
    return (c >= u'A' && c <= u'Z') || (c >= u'a' && c <= u'z');
}

bool StreamXmlPlugin::isTagNameContinuationChar(char16_t c) {
    return isAsciiLetter(c) || (c >= u'0' && c <= u'9') || c == u'_';
}

bool StreamXmlPlugin::isPunctuationTrigger(char16_t c) {
    switch (c) {
        case u'\uFF0C': // ，
        case u'\u3002': // 。
        case u'\uFF1F': // ？
        case u'\uFF01': // ！
        case u'\uFF1A': // ：
        case u'\uFF08': // （
        case u'\uFF09': // ）
        case u'\u3010': // 【
        case u'\u3011': // 】
        case u'\u300A': // 《
        case u'\u300B': // 》
        case u':':
        case u',':
        case u'.':
        case u'?':
        case u'!':
        case u'~':
        case u'\uFF5E': // ～
            return true;
        default:
            return false;
    }
}

bool StreamXmlPlugin::isEmojiTrigger(char16_t c) {
    // Most modern emojis are surrogate pairs in UTF-16.
    if (c >= u'\xD800' && c <= u'\xDFFF') {
        return true;
    }

    // Common BMP emoji/symbol blocks (e.g. ☀, ❤, ✨, etc.).
    if ((c >= u'\x2300' && c <= u'\x23FF') ||
        (c >= u'\x2600' && c <= u'\x27BF') ||
        (c >= u'\x2B00' && c <= u'\x2BFF')) {
        return true;
    }

    return false;
}

bool StreamXmlPlugin::isEmojiContinuationChar(char16_t c) {
    switch (c) {
        case u'\u200D': // ZERO WIDTH JOINER
        case u'\uFE0E': // text presentation selector
        case u'\uFE0F': // emoji presentation selector
        case u'\u20E3': // combining enclosing keycap
            return true;
        default:
            return false;
    }
}

bool StreamXmlPlugin::handleDefaultCharacter(char16_t c) {
    updatePunctuationAllowance(c);
    return true;
}

void StreamXmlPlugin::updatePunctuationAllowance(char16_t c) {
    if (isPunctuationTrigger(c) || isEmojiTrigger(c)) {
        allowStartAfterPunctuation_ = true;
    } else if (c == u' ' || c == u'\t' || isEmojiContinuationChar(c)) {
        // keep
    } else {
        allowStartAfterPunctuation_ = false;
    }
}

bool StreamXmlPlugin::processStartMatcher(char16_t c) {
    switch (startState_) {
        case StartState::WAIT_LT: {
            if (c == u'<') {
                tagName_.clear();
                startState_ = StartState::WAIT_FIRST_LETTER;
                state_ = PluginState::TRYING;
            }
            return false;
        }
        case StartState::WAIT_FIRST_LETTER: {
            if (isAsciiLetter(c)) {
                tagName_.push_back(c);
                startState_ = StartState::IN_TAG_NAME;
                state_ = PluginState::TRYING;
                return false;
            }
            startState_ = StartState::WAIT_LT;
            state_ = PluginState::IDLE;
            return false;
        }
        case StartState::IN_TAG_NAME: {
            if (c == u' ') {
                startState_ = StartState::IN_ATTRS;
                state_ = PluginState::TRYING;
                return false;
            }
            if (c == u'>') {
                startState_ = StartState::WAIT_LT;
                state_ = PluginState::TRYING;
                return true;
            }
            if (!isTagNameContinuationChar(c)) {
                startState_ = StartState::WAIT_LT;
                state_ = PluginState::IDLE;
                tagName_.clear();
                return false;
            }
            tagName_.push_back(c);
            state_ = PluginState::TRYING;
            return false;
        }
        case StartState::IN_ATTRS: {
            if (c == u'>') {
                startState_ = StartState::WAIT_LT;
                state_ = PluginState::TRYING;
                return true;
            }
            state_ = PluginState::TRYING;
            return false;
        }
    }
    return false;
}

void StreamXmlPlugin::buildEndPattern() {
    endPattern_.clear();
    endPattern_.reserve(tagName_.size() + 3);
    endPattern_.push_back(u'<');
    endPattern_.push_back(u'/');
    endPattern_.append(tagName_);
    endPattern_.push_back(u'>');
    endMatcher_.setPattern(endPattern_);
    haveEndPattern_ = true;
}

bool StreamXmlPlugin::processChar(char16_t c, bool atStartOfLine) {
    const char16_t prevChar = lastChar_;
    auto finish = [&](bool result) {
        lastChar_ = c;
        return result;
    };

    if (state_ == PluginState::PROCESSING) {
        if (haveEndPattern_) {
            if (endMatcher_.process(c)) {
                allowStartAfterEndTag_ = true;
                allowStartAfterPunctuation_ = false;
                reset();
                return finish(includeTagsInOutput_);
            }
        }
        return finish(includeTagsInOutput_);
    }

    if (state_ == PluginState::IDLE && !atStartOfLine) {
        const bool allowStart = allowStartAfterEndTag_ || allowStartAfterPunctuation_;
        if (!allowStart) {
            return finish(handleDefaultCharacter(c));
        }
        if (c == u' ' || c == u'\t' || isEmojiContinuationChar(c)) {
            return finish(handleDefaultCharacter(c));
        }
    }

    const PluginState previousState = state_;
    const bool startMatched = processStartMatcher(c);

    if (startMatched) {
        if (prevChar == u'/') {
            // Treat self-closing tags like <br/> as plain text to avoid entering XML mode.
            reset();
            return finish(true);
        }
        state_ = PluginState::PROCESSING;
        allowStartAfterEndTag_ = false;
        allowStartAfterPunctuation_ = false;
        buildEndPattern();
        startState_ = StartState::WAIT_LT;
        return finish(includeTagsInOutput_);
    }

    if (state_ == PluginState::TRYING) {
        allowStartAfterPunctuation_ = false;
        return finish(includeTagsInOutput_);
    }

    if (previousState == PluginState::TRYING) {
        reset();
    }
    allowStartAfterEndTag_ = false;
    allowStartAfterPunctuation_ = false;
    return finish(handleDefaultCharacter(c));
}

} // namespace streamnative

File: android/app/src/main/java/com/ai/assistance/operit/util/stream/plugins/StreamXmlPlugin.kt
package com.ai.assistance.operit.util.stream.plugins

import com.ai.assistance.operit.util.stream.*

private const val GROUP_TAG_NAME = 1
private const val GROUP_CONTENT = 2

/**
 * A stream processing plugin to identify and process XML-formatted data streams. This
 * implementation uses the group capturing mechanism of the StreamKmpGraph. This version has a known
 * limitation: it does not handle nested tags of the same name.
 *
 * @param includeTagsInOutput If true, the XML tags themselves (`<tag>`, `</tag>`) will be included
 * in the output stream. If false, they will be filtered out, leaving only the content between the
 * tags.
 */
class StreamXmlPlugin(private val includeTagsInOutput: Boolean = true) : StreamPlugin {

    override var state: PluginState = PluginState.IDLE
        private set

    private var startTagMatcher: StreamKmpGraph
    private var endTagMatcher: StreamKmpGraph? = null
    // Allow matching a new start tag immediately after we just closed an end tag, even if not at start of line
    private var allowStartAfterEndTag: Boolean = false
    private var allowStartAfterPunctuation: Boolean = false
    private var lastChar: Char = '\u0000'

    private val punctuationTriggers =
            setOf('，', '。', '？', '！', '：', '（', '）', '【', '】', '《', '》', ':', ',', '.', '?', '!', '~', '～')
    private val emojiContinuationChars = setOf('\u200D', '\uFE0E', '\uFE0F', '\u20E3')

    init {
        startTagMatcher =
                StreamKmpGraphBuilder()
                        .build(
                                kmpPattern {
                                    char('<')
                                    // Group 1: Capture the tag name. A valid tag name must start
                                    // with a letter.
                                    // This prevents matching comments (<!--) or closing tags (</).
                                    group(GROUP_TAG_NAME) {
                                        predicate("asciiXmlTagFirstChar") {
                                            it in 'A'..'Z' || it in 'a'..'z'
                                        }
                                        greedyStar {
                                            predicate("xmlTagNameContinuation") {
                                                (it in 'A'..'Z') || (it in 'a'..'z') || (it in '0'..'9') || it == '_'
                                            }
                                        }
                                    }
                                    // Optional: Match attributes until the tag closes
                                    greedyStar { notChar('>') }
                                    char('>')
                                }
                        )

        reset()
    }

    /**
     * Processes a single character for XML stream parsing and decides if it should be emitted. The
     * return value is used by `splitBy` as a filter.
     */
    override fun processChar(c: Char, atStartOfLine: Boolean): Boolean {
        fun finish(result: Boolean): Boolean {
            lastChar = c
            return result
        }

        if (state == PluginState.PROCESSING) {
            // We are inside a tag, looking for the end tag.
            val matcher = endTagMatcher!!
            val result = matcher.processChar(c)

            return when (result) {
                is StreamKmpMatchResult.Match -> {
                    // End tag fully matched. Reset state and filter this last character if needed.
                    StreamLogger.i("StreamXmlPlugin", "Found end tag. Switching to IDLE.")
                    // Enable one-time allowance for starting a new tag right after this end tag
                    allowStartAfterEndTag = true
                    allowStartAfterPunctuation = false
                    reset()
                    finish(includeTagsInOutput)
                }
                is StreamKmpMatchResult.InProgress -> {
                    // We are in the middle of matching the end tag (e.g., '</', '</t', etc.).
                    // The emission of these characters depends on the flag.
                    finish(includeTagsInOutput)
                }
                is StreamKmpMatchResult.NoMatch -> {
                    // The character `c` did not match the next char of the end tag.
                    // This means it's regular content between tags.
                    finish(true)
                }
            }
        } else {
            if (state == PluginState.IDLE && !atStartOfLine) {
                val allowStart = allowStartAfterEndTag || allowStartAfterPunctuation
                if (!allowStart) {
                    return finish(handleDefaultCharacter(c))
                }
                // Allow adjacent XML after an end tag/punctuation even if separated by spaces/tabs
                if (c == ' ' || c == '\t' || isEmojiContinuationChar(c)) {
                    return finish(handleDefaultCharacter(c))
                }
            }
            // We are in IDLE or TRYING state, looking for a start tag.
            val previousState = state
            when (val result = startTagMatcher.processChar(c)) {
                is StreamKmpMatchResult.Match -> {
                    val tagName = result.groups[GROUP_TAG_NAME]
                    if (tagName != null) {
                        if (lastChar == '/') {
                            // Treat self-closing tags like <br/> as plain text to avoid entering XML mode.
                            reset()
                            return finish(true)
                        }
                        StreamLogger.i(
                                "StreamXmlPlugin",
                                "Found start tag '$tagName'. Switching to PROCESSING."
                        )
                        state = PluginState.PROCESSING
                        // Consuming this as a new start clears the post-end allowance
                        allowStartAfterEndTag = false
                        allowStartAfterPunctuation = false
                        // We have a full start tag. Configure the end tag matcher.
                        endTagMatcher =
                                StreamKmpGraphBuilder()
                                        .build(
                                                kmpPattern {
                                                    literal("</")
                                                    literal(tagName)
                                                    char('>')
                                                }
                                        )
                        startTagMatcher.reset()
                    } else {
                        // Should not happen, but as a safeguard:
                        reset()
                    }
                    return finish(includeTagsInOutput)
                }
                is StreamKmpMatchResult.InProgress -> {
                    state = PluginState.TRYING
                    // We are attempting a new start, consume the allowance
                    // so only this potential sequence benefits from it
                    // (if it fails below, we will clear it)
                    // Keep it true while in-progress so subsequent chars can proceed
                    allowStartAfterPunctuation = false
                    return finish(includeTagsInOutput)
                }
                is StreamKmpMatchResult.NoMatch -> {
                    // If we were trying and the match failed, we must reset to idle.
                    if (previousState == PluginState.TRYING) {
                        reset()
                    }
                    // Clear the allowance if we failed to start a new tag
                    allowStartAfterEndTag = false
                    allowStartAfterPunctuation = false
                    // This is a default character, not part of a tag managed by this plugin.
                    return finish(handleDefaultCharacter(c))
                }
            }
        }
    }

    /** Initializes the plugin to its default state. */
    override fun initPlugin(): Boolean {
        reset()
        return true
    }

    /** Destroys the plugin. No-op as listener is removed. */
    override fun destroy() {}

    /** Resets the plugin state. */
    override fun reset() {
        endTagMatcher = null
        startTagMatcher.reset()
        state = PluginState.IDLE
        lastChar = '\u0000'
    }

    private fun handleDefaultCharacter(c: Char): Boolean {
        updatePunctuationAllowance(c)
        return true
    }

    private fun updatePunctuationAllowance(c: Char) {
        when {
            punctuationTriggers.contains(c) || isEmojiTrigger(c) -> {
                allowStartAfterPunctuation = true
            }
            c == ' ' || c == '\t' || isEmojiContinuationChar(c) -> {
                // Keep current state so `<` after spaces/tabs or emoji joiners/selectors still benefits.
            }
            else -> {
                allowStartAfterPunctuation = false
            }
        }
    }

    private fun isEmojiTrigger(c: Char): Boolean {
        // Most modern emojis are surrogate pairs in UTF-16. Treat either half as a trigger.
        if (Character.isSurrogate(c)) {
            return true
        }
        // BMP emoji/symbols (e.g. ☀, ❤) are usually "OTHER_SYMBOL".
        return Character.getType(c) == Character.OTHER_SYMBOL.toInt()
    }

    private fun isEmojiContinuationChar(c: Char): Boolean = emojiContinuationChars.contains(c)
}
